smac/intensification/successive_halving.py

```diff
@@ -0,0 +1,462 @@
+
+        return new_incumbent, inc_perf
+
+    def _top_k(self, configs: typing.List[Configuration], run_history: RunHistory, k: int):
+        """
+        selects the top 'k' configurations from the given list based on their performance in this budget
+
+        Parameters
+        ----------
+        configs: typing.List[Configuration]
+            list of configurations to filter from
+        run_history: RunHistory
+            stores all runs we ran so far
+        k: int
+            number of configurations to select
+
+        Returns
+        -------
+        challengers: typing.List[Configuration]
+            top challenger configurations, sorted in increasing costs
+        """
+        # extracting costs for each given configuration
+        config_costs = {}
+        for c in configs:
+            runs = run_history.get_runs_for_config(c)
+            perf = run_history.aggregate_func(c, run_history, runs)
+            config_costs[c] = perf
+
+        configs_sorted = sorted(config_costs, key=config_costs.get)
+        # select top configurations only
+        top_configs = configs_sorted[:k]
+        return top_configs
+
+    def _adapt_cutoff(self, challenger: Configuration,
+                      incumbent: Configuration,
+                      run_history: RunHistory,
+                      inc_sum_cost: float):
+
+        """Adaptive capping:
+        Compute cutoff based on time so far used for incumbent
+        and reduce cutoff for next run of challenger accordingly
+
+        !Only applicable if self.adaptive capping i.e.,
+            - objective is runtime
+            - sucessive halving is on instances and not cutoff
+            - instance order is not changed per budget
+
+        Parameters
+        ----------
+        challenger : Configuration
+            Configuration which challenges incumbent
+        incumbent : Configuration
+            Best configuration so far
+        run_history : RunHistory
+            Stores all runs we ran so far
+        inc_sum_cost: float
+            Sum of runtimes of all incumbent runs
+
+        Returns
+        -------
+        cutoff: int
+            Adapted cutoff
+        """
+
+        if not self.adaptive_capping:
+            return self.cutoff
+
+        # cost used by challenger for going over all its runs
+        # should be subset of runs of incumbent (not checked for efficiency
+        # reasons)
+        chall_inst_seeds = run_history.get_runs_for_config(challenger)
+        chal_sum_cost = sum_cost(config=challenger,
+                                 instance_seed_pairs=chall_inst_seeds,
+                                 run_history=run_history)
+        cutoff = min(self.cutoff,
+                     inc_sum_cost * self.adaptive_capping_slackfactor -
+                     chal_sum_cost
+                     )
+        return cutoff
+
+    def _compare_configs(self, incumbent: Configuration, challenger: Configuration,
```

Comment:
that's the same function as in the original intensification, right?
if yes, please avoid code duplication. 
(the same applies to the adaptive capping code)